File: src-tauri/src/commands/event.rs
use std::collections::HashMap;
use std::sync::Arc;

use crate::context::ServiceContext;
use log::{debug, error};
use tauri::State;

use wealthfolio_core::events::{Event, EventSpendingSummary, EventWithTypeName};

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateEventInput {
    pub name: String,
    pub description: Option<String>,
    pub event_type_id: String,
    pub start_date: String,
    pub end_date: String,
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateEventInput {
    pub name: Option<String>,
    pub description: Option<String>,
    pub event_type_id: Option<String>,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
}

#[tauri::command]
pub async fn get_events(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<Event>, String> {
    debug!("Fetching all events...");
    state
        .event_service()
        .get_all_events()
        .map_err(|e| {
            error!("Failed to fetch events: {}", e);
            format!("Failed to fetch events: {}", e)
        })
}

#[tauri::command]
pub async fn get_events_with_names(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<EventWithTypeName>, String> {
    debug!("Fetching events with type names...");
    state
        .event_service()
        .get_events_with_type_names()
        .map_err(|e| {
            error!("Failed to fetch events with type names: {}", e);
            format!("Failed to fetch events with type names: {}", e)
        })
}

#[tauri::command]
pub async fn get_event(
    id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Option<Event>, String> {
    debug!("Fetching event: {}", id);
    state
        .event_service()
        .get_event(&id)
        .map_err(|e| {
            error!("Failed to fetch event: {}", e);
            format!("Failed to fetch event: {}", e)
        })
}

#[tauri::command]
pub async fn create_event(
    event: CreateEventInput,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Event, String> {
    debug!("Creating event: {:?}", event.name);
    state
        .event_service()
        .create_event(
            event.name,
            event.description,
            event.event_type_id,
            event.start_date,
            event.end_date,
        )
        .await
        .map_err(|e| {
            error!("Failed to create event: {}", e);
            format!("Failed to create event: {}", e)
        })
}

#[tauri::command]
pub async fn update_event(
    id: String,
    update: UpdateEventInput,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Event, String> {
    debug!("Updating event: {}", id);
    state
        .event_service()
        .update_event(
            &id,
            update.name,
            update.description,
            update.event_type_id,
            update.start_date,
            update.end_date,
        )
        .await
        .map_err(|e| {
            error!("Failed to update event: {}", e);
            format!("Failed to update event: {}", e)
        })
}

#[tauri::command]
pub async fn delete_event(
    event_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<(), String> {
    debug!("Deleting event: {}", event_id);
    state
        .event_service()
        .delete_event(&event_id)
        .await
        .map_err(|e| {
            error!("Failed to delete event: {}", e);
            format!("Failed to delete event: {}", e)
        })?;
    Ok(())
}

#[tauri::command]
pub async fn get_event_activity_counts(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<HashMap<String, i64>, String> {
    debug!("Fetching event activity counts...");
    state
        .event_service()
        .get_activity_counts()
        .map_err(|e| {
            error!("Failed to fetch event activity counts: {}", e);
            format!("Failed to fetch event activity counts: {}", e)
        })
}

#[tauri::command]
pub async fn get_event_spending_summaries(
    start_date: Option<String>,
    end_date: Option<String>,
    base_currency: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<EventSpendingSummary>, String> {
    debug!(
        "Fetching event spending summaries for period: {:?} to {:?}",
        start_date, end_date
    );
    state
        .event_service()
        .get_event_spending_summaries(
            start_date.as_deref(),
            end_date.as_deref(),
            &base_currency,
        )
        .map_err(|e| {
            error!("Failed to fetch event spending summaries: {}", e);
            format!("Failed to fetch event spending summaries: {}", e)
        })
}

File: src/commands/event.ts
import type { Event, EventWithTypeName, NewEvent, UpdateEvent } from "@/lib/types";
import { getRunEnv, RUN_ENV, invokeTauri, invokeWeb } from "@/adapters";
import { logger } from "@/adapters";

export const getEvents = async (): Promise<Event[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_events");
      case RUN_ENV.WEB:
        return invokeWeb("get_events");
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching events.");
    throw error;
  }
};

export const getEventsWithNames = async (): Promise<EventWithTypeName[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_events_with_names");
      case RUN_ENV.WEB:
        return invokeWeb("get_events_with_names");
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching events with names.");
    throw error;
  }
};

export const getEvent = async (id: string): Promise<Event> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_event", { id });
      case RUN_ENV.WEB:
        return invokeWeb("get_event", { id });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching event.");
    throw error;
  }
};

export const createEvent = async (event: NewEvent): Promise<Event> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("create_event", { event });
      case RUN_ENV.WEB:
        return invokeWeb("create_event", { event });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error creating event.");
    throw error;
  }
};

export const updateEvent = async (id: string, update: UpdateEvent): Promise<Event> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("update_event", { id, update });
      case RUN_ENV.WEB:
        return invokeWeb("update_event", { id, update });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error updating event.");
    throw error;
  }
};

export const deleteEvent = async (eventId: string): Promise<void> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        await invokeTauri("delete_event", { eventId });
        return;
      case RUN_ENV.WEB:
        await invokeWeb("delete_event", { eventId });
        return;
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error deleting event.");
    throw error;
  }
};

export const validateTransactionDate = async (
  eventId: string,
  transactionDate: string
): Promise<boolean> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("validate_transaction_date", { eventId, transactionDate });
      case RUN_ENV.WEB:
        return invokeWeb("validate_transaction_date", { eventId, transactionDate });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error validating transaction date.");
    throw error;
  }
};

export const getEventActivityCounts = async (): Promise<Record<string, number>> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_event_activity_counts");
      case RUN_ENV.WEB:
        return invokeWeb("get_event_activity_counts");
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching event activity counts.");
    throw error;
  }
};
